libsrc/apple2/prodos_datetime_to_posix_tm.s

```diff
@@ -0,0 +1,99 @@
+;
+; Oliver Schmidt, 14.08.2018
+; Colin Leroy-Mira, 2023 <colin@colino.net>
+;
+; int __fastcall__ prodos_datetime_to_posix_tm(struct tm *tm, struct prodos_datetime *prodos_dt)
+;
+
+        .import         steaxspidx, incsp2, popax
+        .export         _prodos_datetime_to_posix_tm, prodos_tm_buf
+
+        .include        "time.inc"
+        .include        "zeropage.inc"
+        .include        "errno.inc"
+        .include        "mli.inc"
+
+        ; Convert ProDOS date/time timestamp to POSIX
+        ; source date address in AX
+        ; on stack:
+        ; destination struct
+
+fill_result:
+        jsr     popax
+        cpx     #$00
+        bne     :+
+        cmp     #$00
+        beq     done
+:       sta     ptr1
+        stx     ptr1+1
+        ldy     #$00
+:       lda     prodos_tm_buf,y
+        sta     (ptr1),y
+        iny
+        cpy     #.sizeof(tm)
+        bne     :-
+        rts
+
+zero_tm_buf:
+        ldx     #.sizeof(tm)
+        lda     #$00
+:       dex
+        sta     prodos_tm_buf,x
+        cpx     #$00
+        bne     :-
+        rts
+
+_prodos_datetime_to_posix_tm:
+        sta     ptr1
+        stx     ptr1+1
+
+        ; Get time
+        ldy     #$03
+        lda     (ptr1),y
+        sta     prodos_tm_buf + tm::tm_hour
+        dey
+        lda     (ptr1),y
+        sta     prodos_tm_buf + tm::tm_min
+
+        ; Get date
+        dey
+        lda     (ptr1),y
+        lsr
+        php                     ; Save month msb
+        cmp     #70             ; Year < 70?
+        bcs     :+              ; No, leave alone
+        adc     #100            ; Move 19xx to 20xx
+:       sta     prodos_tm_buf + tm::tm_year
+
+        dey
+        lda     (ptr1),y
+        tax                     ; Save day
+        plp                     ; Restore month msb
+        ror
+        lsr
+        lsr
+        lsr
+        lsr
+        beq     erange          ; [1..12] allows for validity check
+        tay
+        dey                     ; Move [1..12] to [0..11]
+        sty     prodos_tm_buf + tm::tm_mon
+        txa                     ; Restore day
+        and     #%00011111
+        sta     prodos_tm_buf + tm::tm_mday
+
+        ; Copy to destination struct if not NULL
+        jsr     fill_result
+        lda     #$00            ; Return 0
+done:   tax
+        rts
+
+        ; Load errno code
+erange: jsr     zero_tm_buf
+        jsr     fill_result
+        lda     #ERANGE
+        ; Set __errno
+        jmp     ___directerrno
+
+prodos_tm_buf:
+        .tag    tm
```

Comment:
You moved this from .BSS to .CODE - this seems undesirable to me.

Reply:
nice catch!